hal/src/stm32f2xx/i2c_hal.c

```diff
@@ -962,12 +966,18 @@ static void HAL_I2C_EV_InterruptHandler(hal_i2c_interface_t i2c) {
         I2C_StretchClockCmd(i2cMap[i2c]->peripheral, i2cMap[i2c]->clkStretchingEnabled ? ENABLE : DISABLE);
         //I2C_Cmd(i2cMap[i2c]->peripheral, ENABLE);
 
-        i2cMap[i2c]->rxIndexTail = i2cMap[i2c]->rxIndexHead;
-        i2cMap[i2c]->rxIndexHead = 0;
+        if (i2cMap[i2c]->transmitting == 1) { // Slave transmitter
+            i2cMap[i2c]->transmitting = 0;
+        }
 
-        if (NULL != i2cMap[i2c]->callback_onReceive) {
-            // alert user program
-            i2cMap[i2c]->callback_onReceive(i2cMap[i2c]->rxIndexTail);
+        if (i2cMap[i2c]->receiving) { // Slave receiver
+            i2cMap[i2c]->receiving = 0;
```

Comment:
Nit: move into a common function. We have 4 cases where we do exactly the same thing.